File: docs/examples/clusters_dilate.cpp
#include <GooseEYE/GooseEYE.h>
#include <xtensor.hpp>
#include <highfive/H5Easy.hpp>

#define MYASSERT(expr) MYASSERT_IMPL(expr, __FILE__, __LINE__)
#define MYASSERT_IMPL(expr, file, line) \
    if (!(expr)) { \
        throw std::runtime_error( \
            std::string(file) + ':' + std::to_string(line) + \
            ": assertion failed (" #expr ") \n\t"); \
    }

int main()
{
    // generate image
    xt::xarray<int> I = xt::zeros<int>({21, 21});
    I(4, 4) = 1;
    I(14, 15) = 1;
    I(15, 15) = 1;
    I(16, 15) = 1;
    I(15, 14) = 1;
    I(15, 16) = 1;

    // clusters
    auto C = GooseEYE::Clusters(I).labels();

    // dilate
    auto CD = GooseEYE::dilate(C);

    // check against previous versions
    // note that the stored data based unix,
    // the random data is expected to be different on Windows
#ifndef _WIN32
    H5Easy::File data("clusters_dilate.h5", H5Easy::File::ReadOnly);
    MYASSERT(xt::all(xt::equal(I, H5Easy::load<decltype(I)>(data, "I"))));
    MYASSERT(xt::all(xt::equal(C, H5Easy::load<decltype(C)>(data, "C"))));
    MYASSERT(xt::all(xt::equal(CD, H5Easy::load<decltype(CD)>(data, "CD"))));
#endif

    return 0;
}

File: docs/examples/clusters_dilate.py
r'''
    Plot and/or check.

Usage:
    script [options]

Options:
    -s, --save      Save output for later check.
    -c, --check     Check against earlier results.
    -p, --plot      Plot.
    -h, --help      Show this help.
'''

# <snippet>
import numpy as np
import GooseEYE

# generate image
I = np.zeros((21, 21), dtype='bool')
I[4, 4] = True
I[14, 15] = True
I[15, 15] = True
I[16, 15] = True
I[15, 14] = True
I[15, 16] = True

# clusters
C = GooseEYE.Clusters(I).labels()

# dilate
CD = GooseEYE.dilate(C)
# </snippet>

if __name__ == '__main__':

    import docopt

    args = docopt.docopt(__doc__)

    if args['--save']:

        import h5py

        with h5py.File('clusters_dilate.h5', 'w') as data:
            data['I'] = I
            data['C'] = C
            data['CD'] = CD

    if args['--check']:

        import h5py

        with h5py.File('clusters_dilate.h5', 'r') as data:
            assert np.all(np.equal(data['I'][...], I))
            assert np.all(np.equal(data['C'][...], C))
            assert np.all(np.equal(data['CD'][...], CD))

    if args['--plot']:

        import matplotlib.pyplot as plt
        import matplotlib as mpl
        import matplotlib.cm as cm
        from mpl_toolkits.axes_grid1 import make_axes_locatable

        # color-scheme: modify such that the background is white
        # N.B. for a transparent background -> 4th column == 1.
        cmap = cm.jet(range(256))
        cmap[0, :3] = 1.0
        cmap = mpl.colors.ListedColormap(cmap)

        try:
            plt.style.use(['goose', 'goose-latex'])
        except:
            pass

        fig, axes = plt.subplots(figsize=(18, 6), nrows=1, ncols=3)

        ax = axes[0]
        im = ax.imshow(I, clim=(0, 1), cmap=mpl.colors.ListedColormap(cm.gray([0, 255])))
        ax.xaxis.set_ticks([0, 20])
        ax.yaxis.set_ticks([0, 20])
        ax.set_xlim([-0.5, 20.5])
        ax.set_ylim([-0.5, 20.5])
        ax.set_xlabel(r'$x$')
        ax.set_ylabel(r'$y$')
        ax.set_title (r'image')
        div = make_axes_locatable(ax)
        cax = div.append_axes("right", size="5%", pad=0.1)
        cbar = plt.colorbar(im, cax=cax)
        cbar.set_ticks([0, 1])

        ax = axes[1]
        im = ax.imshow(C, clim=(0, np.max(C) + 1), cmap=cmap)
        ax.xaxis.set_ticks([0, 20])
        ax.yaxis.set_ticks([0, 20])
        ax.set_xlim([-0.5, 20.5])
        ax.set_ylim([-0.5, 20.5])
        ax.set_xlabel(r'$x$')
        ax.set_ylabel(r'$y$')
        ax.set_title (r'clusters')

        ax = axes[2]
        im = ax.imshow(CD, clim=(0, np.max(C) + 1), cmap=cmap)
        ax.xaxis.set_ticks([0, 20])
        ax.yaxis.set_ticks([0, 20])
        ax.set_xlim([-0.5, 20.5])
        ax.set_ylim([-0.5, 20.5])
        ax.set_xlabel(r'$x$')
        ax.set_ylabel(r'$y$')
        ax.set_title (r'clusters + dilate')

        plt.savefig('clusters_dilate.svg')
